### ISSUE TITLE
Test database not reset between tests

### ISSUE THREAD

**Issue by nilnullzip** [mad-eye/meteor-mocha-web#136]
Is this the expected behavior for Mocha/Velocity?

**Comment by rissem** [comment]
That is expected right now. We should probably do a better job communicating that. It sounds like you would like to have the database cleared between tests? Or between test runs?

**Comment by nilnullzip** [comment]
es, the database needs to be cleared to ensure that the starting conditions are the same every time. Otherwise you can get unrepeatable results. Jasmine appears to clear the database as my app runs through its initialization and sets up the database from the empty state.

At the very least it needs to be cleared between test runs. I'm not sure I understand the exact distinction between tests and test runs. So I'm not sure how to answer the question if it has to be cleared between tests too.

**Comment by rissem** [comment]
I agree that some kind of reset should happen. The question to me is who should be responsible for resetting (framework, Velocity, or user), and how often it should happen. Resetting the db after each test seems like it'd be too slow, but at the start of the test suite seems like it'd offer a good balance of speed and a clear starting point.

**Comment by nilnullzip** [comment]
I would think the logical place is in Velocity when the mirror is set up. Does that happen only once at the beginning of a suite of tests? Or does it happen before every test?

Actually I think clearing the database should be very fast in Mongo as it is accomplished by deleting all of the DB files.

**Comment by aramk** [comment]
Yeah I agree, following discussion from https://github.com/meteor-velocity/velocity/issues/39 it seems like this should already be implemented in the mirror? In the meantime, I'm using a utility method I wrote [here](https://github.com/aramk/meteor-utility/blob/master/src/Collections.coffee#L262).

```
beforeEach ->
      Collections.removeAllDocs(Orders)
      Collections.removeAllDocs(OrderSchedules)
```

**Comment by aramk** [comment]
I've found a slightly better temporary solution. If you have a `beforeEach` outside of any `describe` in Mocha, it will run for each test case in all suites, so I created a test called `_startupTest.coffee` - I'm assuming it will load tests in alphabetical order.

```
MochaWeb?.testOnly ->

  beforeEach ->
    console.log('Clearing test collections')
    _.each [Orders, OrderSchedules, Runs, RunSchedules, Customers, Products], (collection) ->
      Collections.removeAllDocs(collection)
```

**Comment by OminStyle** [comment]
Is there a meteor method do get this done from Velocity?

**Comment by carlosbaraza** [comment]
:+1: Any news about this? How do you guys usually clear the DB after tests without modifying the actual mocha package?

**Comment by OminStyle** [comment]
@carlosbaraza Although in my opinion this is not the best solution, what I ended up doing is remove every item in each collection manually using beforeEach(). I did this because my current permissions do not allow me to drop databases on the client side.

```
if (!(typeof MochaWeb === 'undefined')){
  MochaWeb.testOnly(function(){
    beforeEach(function() {
      Collection1.find().forEach(function(item) {
        Collection1.remove(item._id);
      });
      Collection2.find().forEach(function(item) {
        Collection2.remove(item._id);
      });
    });
  });
}
```

**Comment by carlosbaraza** [comment]
Hi @OminStyle,

But what about cleaning the full database, not only the documents available in the client? I want to have a clean way to clear the database without compromising security, letting docs to be deleted from the client.

Maybe I am missing still some point from Velocity. Is there some way to set up a server method only callable from tests where I could clear the database?

**Comment by ryneeverett** [comment]
@carlosbaraza Agree that this isn't ideal, but couldn't you simply add the methods in a `debugOnly` package?

**Comment by carlosbaraza** [comment]
Hi @ryneeverett, thank you, that's an interesting point. I am still kind of new on the Meteor world, so I am figuring out the best way to do things. That might be my solution for the moment.

Do you have some example repository where this is being done?

**Comment by carlosbaraza** [comment]
For the moment I could not find someone doing such a debug package to clear the database via methods. However, I found a cool well known package example of the usage of `debugOnly` https://github.com/msavin/Mongol/blob/master/package.js

**Comment by nilnullzip** [comment]
Try this one for clearing the database:

> https://github.com/Sanjo/meteor-jasmine/wiki/Database-Fixtures-for-Integration-Tests

**Comment by ryneeverett** [comment]
Here's my mocha implementation:

https://github.com/ryneeverett/meteor-dbreset

**Comment by Valentin-Seehausen** [comment]
And here is my implementation. This will clear all Items from Collections published to Client, If Client has permissions for deleting them. Save this in client test folder. Name could be 00.clearDatabaseTest.coffee, to assure initial load.

	MochaWeb?.testOnly ->

		collections = [Collection1, Collection2]
		clearDatabase = ->
			console.log('Clearing test collections')
			_.each collections, (collection) ->
				_.each collection.find().fetch(), (doc) ->
					doc.remove()


		describe 'Clearing Database', ->
			it 'should remove all items from Collections', ->
				clearDatabase()
				_.each collections, (collection) ->
					chai.assert.equal collection.find().count(), 0
